pkg/controller/runner.go

```diff
@@ -4,34 +4,32 @@ import (
 	"context"
 
 	"github.com/go-logr/logr"
-
-	"github.com/aws/eks-anywhere/pkg/cluster"
 )
 
 // Phase represents a generic reconciliation phase for a cluster spec.
-type Phase func(ctx context.Context, log logr.Logger, clusterSpec *cluster.Spec) (Result, error)
+type Phase[O any] func(ctx context.Context, log logr.Logger, obj O) (Result, error)
 
 // PhaseRunner allows to execute Phases in order.
-type PhaseRunner struct {
-	phases []Phase
+type PhaseRunner[O any] struct {
+	phases []Phase[O]
```

Comment:
I'm not opposed. Have we considered another approach where we simply use add state to the PhaseRunner struct?

E.g.

```
type PhaseRunner struct {
  phases []Phase
  isRollingUpdate: bool
}
```

Reply:
This would add isRollingUpdate to phase runner in all providers, but it's just tinkerbell specific

Reply:
This may be good to keep as an alternative, I think Chris saw some code smells with how I'm using the generic phase runner. I ended up still wrapping the existing IPValidatorPhase which takes in clusterSpec as an argument with a Tinkerbell phase just so I could run it, which might be moving backwards. 

Maybe we an additional field called additionalDependencies or something could do what you're suggesting.

```
type PhaseRunner struct {
  phases []Phase
 additionalDependencies  any
}

type TinkerbellAdditionalDeps struct {
  isRollingUpdate  bool
  controlPlane   CP
  workerNodes  W
}
```

Reply:
Yeah, I realize it's not exactly provider specific, but the generic does seem like an overkill solution.

@raymond-zhang00 Yeah. I would even make that a `map[string]interface{}`, if we decide to go with this as an alternative.

Reply:
Ended up having a discussing the code smell with Chris and utilizing a generic phase runner should be a solid approach so I don't think we'll need to go down this path.

Reply:
Okay, gotcha!

Reply:
Was this based on a design that @g-gaston shared recently?

Reply:
Yes, we discussed that we wanted an object to hold data from phase to phase that's thread safe but didn't want to add on to ClusterSpec.